File: Algorithms/1570_Dot_Product_of_Two_Sparse_Vectors/sparse_vector.rb
class SparseVector
  attr_reader :nums

=begin
    :type nums: Integer[]
=end
  def initialize(nums)
    @nums = nums
  end

  # Return the dotProduct of two sparse vectors
=begin
    :type vec: SparseVector
    :rtype: Integer
=end
  def dotProduct(vec)
    acc = 0
    @nums.each_with_index { |v, i| acc += v * vec.nums[i] }
    acc
  end
end

# Your SparseVector object will be instantiated and called as such:
# v1 = SparseVector.new(nums1)
# v2 = SparseVector.new(nums2)
# ans = v1.dotProduct(v2)

File: Algorithms/1570_Dot_Product_of_Two_Sparse_Vectors/sparsevector.go
package main

type SparseVector struct {
	nums []int
}

func Constructor(nums []int) SparseVector {
	return SparseVector{nums}
}

// Return the dotProduct of two sparse vectors
func (sv *SparseVector) dotProduct(vec SparseVector) int {
	acc := 0
	for i, v := range sv.nums {
		acc += v * vec.nums[i]
	}
	return acc
}

/**
 * Your SparseVector object will be instantiated and called as such:
 * v1 := Constructor(nums1);
 * v2 := Constructor(nums2);
 * ans := v1.dotProduct(v2);
 */
